// igniteui-angular/esm2022/lib/animations/grow/index.mjs
import { animate, animation, style } from '@angular/animations';
import { EaseIn, EaseOut } from '../easings';
const base = [
    style({
        opacity: `{{ startOpacity }}`,
        height: `{{ startHeight }}`,
        paddingBlock: `{{ startPadding }}`
    }),
    animate(`{{duration}} {{delay}} {{easing}}`, style({
        opacity: `{{ endOpacity }}`,
        height: `{{ endHeight }}`,
        paddingBlock: `{{ endPadding }}`
    }))
];
const baseParams = {
    delay: '0s',
    duration: '350ms',
    easing: EaseIn.Quad,
    startOpacity: 0,
    endOpacity: 1,
    startHeight: '',
    endHeight: '',
    startPadding: '',
    endPadding: '',
};
const growVerIn = animation(base, {
    params: {
        ...baseParams,
        easing: EaseOut.Quad,
        startOpacity: 0,
        endOpacity: 1,
        startHeight: '0px',
        endHeight: '*',
        startPadding: '0px',
        endPadding: '*'
    }
});
const growVerOut = animation(base, {
    params: {
        ...baseParams,
        easing: EaseOut.Quad,
        startOpacity: 1,
        endOpacity: 0,
        startHeight: '*',
        endHeight: '0px',
        startPadding: '*',
        endPadding: '0px'
    }
});
export { growVerIn, growVerOut };
