// igniteui-angular/esm2022/lib/tabs/tab-item.directive.mjs
import { ContentChild, Directive, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { Direction } from '../carousel/carousel-base';
import { IgxTabHeaderBase, IgxTabContentBase } from './tabs.base';
import * as i0 from "@angular/core";
import * as i1 from "./tabs.base";
export class IgxTabItemDirective {
    /**
     * An @Input property which determines whether an item is selected.
     */
    get selected() {
        return this._selected;
    }
    set selected(value) {
        if (this._selected !== value) {
            this._selected = value;
            this.tabs.selectTab(this, this._selected);
            this.selectedChange.emit(this._selected);
        }
    }
    /** @hidden */
    constructor(tabs) {
        this.tabs = tabs;
        /**
         * Output to enable support for two-way binding on [(selected)]
         */
        this.selectedChange = new EventEmitter();
        /**
         * An @Input property that allows you to enable/disable the item.
         */
        this.disabled = false;
        /** @hidden */
        this.direction = Direction.NONE;
        this._selected = false;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.1.2", ngImport: i0, type: IgxTabItemDirective, deps: [{ token: i1.IgxTabsBase }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.1.2", type: IgxTabItemDirective, inputs: { disabled: "disabled", selected: "selected" }, outputs: { selectedChange: "selectedChange" }, queries: [{ propertyName: "headerComponent", first: true, predicate: IgxTabHeaderBase, descendants: true }, { propertyName: "panelComponent", first: true, predicate: IgxTabContentBase, descendants: true }], viewQueries: [{ propertyName: "headerTemplate", first: true, predicate: ["headerTemplate"], descendants: true, static: true }, { propertyName: "panelTemplate", first: true, predicate: ["panelTemplate"], descendants: true, static: true }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.1.2", ngImport: i0, type: IgxTabItemDirective, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i1.IgxTabsBase }]; }, propDecorators: { headerComponent: [{
                type: ContentChild,
                args: [IgxTabHeaderBase]
            }], panelComponent: [{
                type: ContentChild,
                args: [IgxTabContentBase]
            }], headerTemplate: [{
                type: ViewChild,
                args: ['headerTemplate', { static: true }]
            }], panelTemplate: [{
                type: ViewChild,
                args: ['panelTemplate', { static: true }]
            }], selectedChange: [{
                type: Output
            }], disabled: [{
                type: Input
            }], selected: [{
                type: Input
            }] } });
